nixos/modules/services/mail/cyrus-imap.nix

```diff
@@ -0,0 +1,366 @@
+            default = "/run/cyrus/idle";
+            description = ''
+              Unix socket that idled listens on.
+            '';
+          };
+          notifysocket = mkOption {
+            type = path;
+            default = "/run/cyrus/notify";
+            description = ''
+              Unix domain socket that the mail notification daemon listens on.
+            '';
+          };
+        };
+      };
+      default = {
+        admins = [ "cyrus" ];
+        allowplaintext = true;
+        defaultdomain = "localhost";
+        defaultpartition = "default";
+        duplicate_db_path = "/run/cyrus/db/deliver.db";
+        hashimapspool = true;
+        httpmodules = [
+          "carddav"
+          "caldav"
+        ];
+        mboxname_lockpath = "/run/cyrus/lock";
+        partition-default = "/var/lib/cyrus/storage";
+        popminpoll = 1;
+        proc_path = "/run/cyrus/proc";
+        ptscache_db_path = "/run/cyrus/db/ptscache.db";
+        sasl_auto_transition = true;
+        sasl_pwcheck_method = [ "saslauthd" ];
+        sievedir = "/var/lib/cyrus/sieve";
+        statuscache_db_path = "/run/cyrus/db/statuscache.db";
+        syslog_prefix = "cyrus";
+        tls_client_ca_dir = "/etc/ssl/certs";
+        tls_session_timeout = 1440;
+        tls_sessions_db_path = "/run/cyrus/db/tls_sessions.db";
+        virtdomains = "on";
+      };
+      description = "IMAP configuration settings. [imapd.conf(5)](https://www.cyrusimap.org/imap/reference/manpages/configs/imapd.conf.html)";
+    };
+
+    user = mkOption {
+      type = str;
+      default = "cyrus";
+      description = "Cyrus IMAP user name.";
+    };
+
+    group = mkOption {
+      type = str;
+      default = "cyrus";
+      description = "Cyrus IMAP group name.";
+    };
+
+    imapdConfigFile = mkOption {
+      type = nullOr path;
+      default = null;
+      description = "Config file used for the whole cyrus-imap configuration.";
+      apply = v: if v != null then v else pkgs.writeText "imapd.conf" imapdConfig;
+    };
+
+    cyrusConfigFile = mkOption {
+      type = nullOr path;
+      default = null;
+      description = "Config file used for the whole cyrus-imap configuration.";
+      apply = v: if v != null then v else pkgs.writeText "cyrus.conf" cyrusConfig;
+    };
+
+    sslCACert = mkOption {
+      type = nullOr str;
+      default = null;
+      description = "File containing one or more Certificate Authority (CA) certificates.";
+    };
+
+    sslServerCert = mkOption {
+      type = nullOr str;
+      default = null;
+      description = "File containing the global certificate used for ALL services (imap, pop3, lmtp, sieve)";
+    };
+
+    sslServerKey = mkOption {
+      type = nullOr str;
+      default = null;
+      description = "File containing the private key belonging to the global server certificate.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.users.cyrus = optionalAttrs (cfg.user == "cyrus") {
```

Comment:
Is a user just not created if `cfg.user` is set to something else? That should be documented.

Reply:
Isn’t it a common practice in NixOS modules?

Reply:
Common practice seems to be a bit all over the place. For example, `appvm` creates the user with the name that's been given, `zitadel` does the same thing as you do here and only creates the user and group if the name is set to `zitadel`, and `tomcat` unconditionally creates a user and group both named `tomcat`.

Personally, I still think the implicit behavior of creating a user/group like this kind of weird, and I think either one of these two options could be viable:
- Default the user and group options to `null`, and the default user and group are only created of the options are null;
- Set the default user and group with the given name, but make it `mkDefault` so that downstream users can override them if they want to.

Reply:
Well…
I‘d like to set `null` for `user` and `group`.